Hadisd simplify 146

millerjoel opened the pull request:
Description:
- Improve notebooks 1 & 2 so they should now be able to be run at any stage and repeat downloads won't happen
- Remove need for station ranges in file system method
- Improve efficiency of converting to Zarr using dask delayed

Testing:
- Worth checking that notebooks 1,2 & 3 work from start to finish. Should be able to interrupt 1 & 2 and they should start where they left off.

millerjoel committed:
Change get_all_station_ids method to work with new zarr structure

millerjoel committed:
Update filesystem method to work with simplified file structure

millerjoel committed:
Add functionality to loop over station rnges for downloads

millerjoel committed:
Make downloading and extraction even more idempotent. Final cell of downloads can be run as many times as we like

millerjoel committed:
Improve conversion efficency

millerjoel committed:
Tidy station list

millerjoel committed:
Improve notebook descriptions

tennlee commented:
I have done a lot of behind-the-scenes work on the static analysis on develop. Please rebase your branch, and ideally also make it compliant with 'ruff check'

Commit:
Ruff compliance added

millerjoel commented:
I've made sure my changes are technically ruff compliant. Some notebooks fail ruff checks due to run magic & defining variable names in a separate notebook. I've told ruff to ignore this, specifically for the notebooks that cause the problem. Before hitting the merge button, I'll bring this up in our meeting tomorrow.

tennlee commented:
I'm just working through this. The download notebook worked well, looks good to me. The conversion to Zarr issues a lot of warnings. It would be nice if it quieted the warnings after the first one.

tennlee commented:
The tutorial depends on seaborn, which isn't currently part of the dependencies. I don't think PET wants to influence people's choice of plotting library, there are lots of sensible options and people should feel free to use whatever. I don't think it has many dependencies so I'm happy to add it, but you also might want to just consider if you want to redo the plotting with matplotlib or plotly. So no issue - but please confirm your choice, and add it to the dependencies if it's needed. I'll just manually install it for my testing purposes for now.

tennlee commented:
I'm getting an error which seems to be related to the warning message I saw when converting to Zarr. The warning message is "ValueError: Invalid Zarr format 3 data_type: {'name': 'fixed_length_utf32', 'configuration': {'length_bytes': 48}}". Did you get that at any stage?

millerjoel committed:
Merge branch 'ACCESS-Community-Hub:develop' into hadisd-simplify-146

millerjoel committed:
Update Drop transform for more robust handling

millerjoel committed:
Update download notebook to show good output

millerjoel committed:
Update conversion to zarr notebook to show good output

millerjoel committed:
Add docstrings

millerjoel commented:
I've made some changes to reflect your comments:
1) Zarr warnings no longer seem to be an issue, so have removed the code that suppresses them

2) Plotting is now done with matplotlib only, no seaborn.

In the mean time I seem to have broken something whilst looking into why the Drop transform was getting rid of coordinates I didn't want it to. I've reverted back to the way I thought things were, but there is still a problem I need to look into.

tennlee commented:
Thanks I'll take another look shortly.

millerjoel committed:
Merge branch 'ACCESS-Community-Hub:develop' into hadisd-simplify-146

tennlee committed:
Improve error message for numpy conversion process

millerjoel committed:
Run again with good output after variable/numpy fix

millerjoel committed:
Update pre-processing to remove redundant station_id variable causing numpy issues, also squeeze lat, lon and elev here instead of pipeline

millerjoel committed:
Update pre-processing to remove redundant station_id variable causing numpy issues, also squeeze lat, lon and elev here instead of pipeline

millerjoel committed:
Update notebook to plot confusion matrix with matplotlib and adjust pipeline

millerjoel commented:
Right, should all be working again. Thanks again for your help!

I've adjusted variables.Drop so it doesn't drop coordinates that I haven't asked it to. I did start thinking this might be intended behaviour, but I don't really think it's that useful. This meant that I did have to manually drop some coordinates (lat, lon, elev) as part of the pipeline, but I think it's better that I explicitly have to drop them. This was done so that .ToNumpy() would work correctly. 

The other numpy problem was that there was a station_id variable that only had a station dimension, no time. I have just dropped it during the preprocessing stage in notebook 2 since it's a bit redundant anyway. This leaves what I think is quite a sensible dataset.

Perhaps give everything one more run to check it works and if it does then I'll merge it.

I'll then have another PR soon that addresses the zarr chunking issues experience on HPCs.

tennlee commented:
This is also working for me now. I'll merge this increment in, but I can see some possible improvements for another day. Thanks very much for the work!

tennlee commented on packages/data/src/pyearthtools/data/transforms/variables.py:
```diff
@@ -86,12 +86,7 @@ def __init__(self, variables: list[str] | str, *extra_variables):
     def apply(self, dataset: xr.Dataset) -> xr.Dataset:
         if self._variables is None:
             return dataset
-
```
I'm not really sure why the original code was doing what it's doing. It seem like it's doing a double-negative instead of just dropping what as asked. There's no code comments in the original about why it is doing things that way, and I think most people would be totally comfortable with a transform that just does the drop without complex checking. I'll get in there and add back the old code in a comments block with some notes around it in case we need to come back to it, and then if it's not important after 6 months or so we can delete it.

tennlee replied:
Ah, I think it's trying to avoid dropping coordinate variables. So I think if you pass a coordinate variable name in, it won't drop it. So perhaps it would be better off called "DropDataVariable" and have another one for dropping coordinates. It might be worth either making two classes with a clear purpose, or else raising a warning if dropping a coordinate variable. Obviously people do need a way to drop coordinate variables when needed, but it's riskier because you might not be able to index into your data any more...

tennlee committed:
Skip radar projection tests if radar data is unavailable
Put in some commentary about the dropping of coordinates with the Drop operator

tennlee committed:
Run code reformatting

tennlee committed:
Address code linting tool feedback
